spokestack/asr/cloud_client.py

```diff
@@ -0,0 +1,106 @@
+"""
+This module contains the websocket logic used to communicate with
+Spokestack's cloud-based ASR service.
+"""
+import base64
+import hashlib
+import hmac
+import json
+
+from websocket import WebSocket  # type: ignore
+
+
+class CloudClient:
+    """ spokestack cloud client """
+
+    def __init__(
+        self,
+        socket_url: str,
+        key_id: str,
+        key_secret: str,
+        audio_format: str = "PCM16LE",
+        sample_rate: int = 16000,
+        language: str = "en",
+        limit: int = 10,
+    ) -> None:
+        self._body: str = json.dumps(
+            {
+                "format": audio_format,
+                "rate": sample_rate,
+                "language": language,
+                "limit": limit,
+            }
+        )
+
+        self._socket_url: str = socket_url
+        self._key_id: str = key_id
+        self._key: bytes = key_secret.encode("utf-8")
+        signature = hmac.new(
+            self._key, self._body.encode("utf-8"), hashlib.sha256
+        ).digest()
+        self._signature = base64.b64encode(signature).decode("utf-8")
+        self._socket: WebSocket = WebSocket()
+        self._response: dict = {
+            "error": None,
+            "final": False,
+            "hypotheses": [],
+            "status": None,
+        }
+        self._connected = False
+        self._is_final = False
+
+    @property
+    def is_connected(self) -> bool:
+        return self._connected
+
+    def close(self) -> None:
+        self._socket.close()
+
+    def connect(self) -> None:
+        if self._connected:
+            raise ConnectionError("Already Connected")
+        self._socket.connect(f"{self._socket_url}/v1/asr/websocket")
+        self._connected = True
+
+    def initialize(self) -> None:
+        if not self._connected:
+            raise ConnectionError("Not Connected")
+
+        message = {
+            "keyId": self._key_id,
+            "signature": self._signature,
+            "body": self._body,
+        }
+        self._socket.send(json.dumps(message))
+        self._socket.recv()
```

Comment:
it would probably be better to convert api error responses to exceptions here (and in the receive function), in order to avoid silent failure for applications
I would create a new exception class and use it to encapsulate failed responses